unittests: Fixes unit tests for Windows (part 2)

claudiubelu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind failing-test

/sig testing
/sig windows

#### What this PR does / why we need it:

Currently, there are some unit tests that are failing on Windows due to various reasons:

- volume mounting is a bit different on Windows: Mount will create the parent dirs and mklink at the volume path later (otherwise mklink will raise an error).
- ``os.Chmod`` is not working as intended on Windows.
- ``path.Dir()`` will always return ``"."`` on Windows, and ``filepath.Dir()`` should be used instead (which works correctly).
- on Windows, you can't typically run binaries without extensions. If the file ``C:\\foo.bat`` exists, we can still run ``C:\\foo`` because Windows will append one of the supported file extensions (``$env:PATHEXT``) to it and run it.
- Windows file permissions do not work the same way as the Linux ones.
- ``/tmp`` directory being used, which might not exist on Windows. Instead, the OS-specific Temp directory should be used.

Fixes a few other issues:

- ``rbd.go``: Return error in a case in which an error is encountered. This will prevent ``"rbd: failed to setup"`` and ``"rbd: successfully setup"`` log messages to be logged at the same time.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related: #51540

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

claudiubelu committed:
unittests: Fixes unit tests for Windows (part 2)

Currently, there are some unit tests that are failing on Windows due to
various reasons:

- volume mounting is a bit different on Windows: Mount will create the
  parent dirs and mklink at the volume path later (otherwise mklink will
  raise an error).
- os.Chmod is not working as intended on Windows.
- path.Dir() will always return "." on Windows, and filepath.Dir()
  should be used instead (which works correctly).
- on Windows, you can't typically run binaries without extensions. If
  the file C:\\foo.bat exists, we can still run C:\\foo because Windows
  will append one of the supported file extensions ($env:PATHEXT) to it
  and run it.
- Windows file permissions do not work the same way as the Linux ones.
- /tmp directory being used, which might not exist on Windows. Instead,
  the OS-specific Temp directory should be used.

Fixes a few other issues:

- rbd.go: Return error in a case in which an error is encountered. This
  will prevent "rbd: failed to setup" and "rbd: successfully setup" log
  messages to be logged at the same time.

claudiubelu commented:
/retest

fedebongio commented:
/remove-sig api-machinery

claudiubelu commented:
/milestone v1.25
/priority important-soon

claudiubelu commented:
/test pull-kubernetes-e2e-capz-windows-containerd

claudiubelu commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

jsturtevant commented:
/triage accepted

claudiubelu commented:
/assign @thockin

thockin commented on pkg/volume/csi/csi_attacher_test.go:
```diff
@@ -1216,7 +1217,10 @@ func TestAttacherMountDevice(t *testing.T) {
 			populateDeviceMountPath: true,
 			shouldFail:              true,
-			spec:                    volume.NewSpecFromPersistentVolume(makeTestPV(pvName, 10, testDriver, "test-vol1"), true),
+			// NOTE: We're skipping this test on Windows because os.Chmod is not working as intended, which means that
```
is this temporary or "it doesn't work that way" ?

claudiubelu replied:
Windows permissions work a bit differently, and are a bit more fine grained. However, the Windows permissions could be boiled down to something Linux-like. I do have a PR for something like this (https://github.com/kubernetes/kubernetes/pull/104660), and we could revisit this particular skip if that gets merged.

thockin replied:
So are we OK to merge this without Windows supporting it?

@msau42 @jingxu97

jingxu97 replied:
I think it is ok to add a TODO and revisit here after windows permission support PR is merged.

Also want to confirm if in test we do not use Chmod to change file permission, the test will fail for Linux?

claudiubelu replied:
Added TODO. Also, the test will fail on Linux without the Chmods:

```
=== RUN   TestAttacherMountDevice/failure_PV_with_existing_data
    csi_attacher_test.go:1294: Running test case: failure PV with existing data
    csi_client_test.go:280: calling fake.NodeGetCapabilities for NodeSupportsStageUnstage...
    csi_client_test.go:227: calling fake.NodeStageVolume...
    csi_attacher_test.go:1369: test should fail, but no error occurred
```

markyjackson-taulia commented:
Hi @claudiubelu  My name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Just checking in to see if this is still on track for k8s 1.25?

claudiubelu commented:
> Hi @claudiubelu  My name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Just checking in to see if this is still on track for k8s 1.25?

Hello. Yes, we'd like to have this merged in v1.25.

claudiubelu commented:
/sig testing

jingxu97 commented on pkg/volume/awsebs/aws_ebs_block.go:
```diff
@@ -66,6 +66,8 @@ func (plugin *awsElasticBlockStorePlugin) getVolumeSpecFromGlobalMapPath(volumeN
 	fullVolumeID := strings.TrimPrefix(globalMapPath, pluginDir) // /vol-XXXXXX
 	fullVolumeID = strings.TrimLeft(fullVolumeID, "/")           // vol-XXXXXX
+	// Windows paths have \\ instead.
+	fullVolumeID = strings.TrimLeft(fullVolumeID, "\\") // vol-XXXXXX
```
do we need to check whether it is windows os?

claudiubelu replied:
Hm, we could check... although it shouldn't be necessary, unless the ``fullVolumeId`` would have some leading backslashes for some reason. But better safe than sorry.

claudiubelu commented:
/retest

claudiubelu commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

claudiubelu commented:
/retest

claudiubelu commented:
/retest

claudiubelu commented:
/retest

thockin commented:
Is this ready for review?

claudiubelu commented:
> Is this ready for review?

Yes.

thockin commented on pkg/volume/configmap/configmap_test.go:
```diff
@@ -361,7 +361,9 @@ func TestPlugin(t *testing.T) {
 
 	volumePath := mounter.GetPath()
-	if !strings.HasSuffix(volumePath, fmt.Sprintf("pods/test_pod_uid/volumes/kubernetes.io~configmap/test_volume_name")) {
+	expectedSuffix := "pods/test_pod_uid/volumes/kubernetes.io~configmap/test_volume_name"
```
Why not capture this like:

```
func hasPathSuffix(s, suffix) bool {
	// Convert to a platform-native path.
	sfx := strings.ReplaceAll(in, "/", string(os.PathSeparator))
	return strings.HasSuffix(s, sfx)
}
```

and then each callsite is just:

```
if !hasPathSuffix(volumePath, "pods/test_pod_uid/volumes/kubernetes.io~configmap/test_volume_name") {
```

(note `ReplaceAll`, I think linters will warn)

thockin commented on pkg/volume/csi/csi_mounter_test.go:
```diff
@@ -905,19 +905,25 @@ func TestUnmounterTeardown(t *testing.T) {
 	// save the data file prior to unmount
-	dir := filepath.Join(getTargetPath(testPodUID, pv.ObjectMeta.Name, plug.host), "/mount")
+	targetDir := getTargetPath(testPodUID, pv.ObjectMeta.Name, plug.host)
+	dir := filepath.Join(targetDir, "/mount")
```
I think you don't want the '/' on the last arg?

thockin commented on pkg/volume/csi/csi_util_test.go:
```diff
@@ -116,12 +115,13 @@ func TestSaveVolumeData(t *testing.T) {
 		specVolID := fmt.Sprintf("spec-volid-%d", i)
-		mountDir := filepath.Join(getTargetPath(testPodUID, specVolID, plug.host), "/mount")
+		targetPath := getTargetPath(testPodUID, specVolID, plug.host)
+		mountDir := filepath.Join(targetPath, "/mount")
```
I think you don't want the '/' on the last arg?

claudiubelu replied:
Done

thockin commented on pkg/volume/gcepd/attacher_test.go:
```diff
@@ -96,8 +97,10 @@ func TestAttachDetachRegional(t *testing.T) {
-			if devicePath != "/dev/disk/by-id/google-disk" {
-				return fmt.Errorf("devicePath incorrect. Expected<\"/dev/disk/by-id/google-disk\"> Actual: <%q>", devicePath)
+			expectedDevicePath := "/dev/disk/by-id/google-disk"
+			expectedDevicePath = strings.Replace(expectedDevicePath, "/", string(os.PathSeparator), -1)
```
Same as above - capturing this to a function would make it clearer

thockin commented on pkg/volume/git_repo/git_repo_test.go:
```diff
@@ -314,7 +314,7 @@ func doTestPlugin(scenario struct {
 
 	path := mounter.GetPath()
-	suffix := fmt.Sprintf("pods/poduid/volumes/kubernetes.io~git-repo/%v", scenario.vol.Name)
+	suffix := filepath.Join("pods/poduid/volumes/kubernetes.io~git-repo/", scenario.vol.Name)
```
don't want trailing '/' ?

thockin commented on pkg/volume/rbd/rbd.go:
```diff
@@ -44,6 +44,7 @@ import (
 
 var (
 	supportedFeatures = sets.NewString("layering")
+	pathSeparator     = string(os.PathSeparator)
```
why this indirection?

thockin commented on pkg/volume/secret/secret_test.go:
```diff
@@ -323,7 +323,8 @@ func TestPlugin(t *testing.T) {
 
 	volumePath := mounter.GetPath()
-	if !strings.HasSuffix(volumePath, fmt.Sprintf("pods/test_pod_uid/volumes/kubernetes.io~secret/test_volume_name")) {
+	expectedSuffix := strings.Replace("pods/test_pod_uid/volumes/kubernetes.io~secret/test_volume_name", "/", string(os.PathSeparator), -1)
```
function?

It doesn't haver to be in some util pkg, just in each test

thockin reviewed (commented):
small things

adelina-t commented:
besides minor nits that @thockin captured: /lgtm

claudiubelu commented on pkg/volume/cinder/attacher_test.go:
```diff
@@ -89,7 +91,7 @@ func TestGetDeviceMountPath(t *testing.T) {
 		t.Errorf("Get device mount path error")
 	}
-	expectedPath := rootDir + "plugins/kubernetes.io/cinder/mounts/" + name
+	expectedPath := filepath.Join(rootDir, "plugins/kubernetes.io/cinder/mounts/", name)
```
True. ``filepath.Join`` will also clean the resulting path, so there are no duplicate slashes, but it would be cleaner to remove the ``/``.

thockin replied:
It looks like filepath.Clean does handle this and is documented

claudiubelu commented on pkg/volume/cinder/attacher_test.go:
```diff
@@ -89,7 +91,7 @@ func TestGetDeviceMountPath(t *testing.T) {
 		t.Errorf("Get device mount path error")
 	}
-	expectedPath := rootDir + "plugins/kubernetes.io/cinder/mounts/" + name
+	expectedPath := filepath.Join(rootDir, "plugins/kubernetes.io/cinder/mounts/", name)
```
Done

claudiubelu commented:
/test pull-kubernetes-unit

markyjackson-taulia commented:
Hello @claudiubelu, my name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Is this still on track for 1.25?

claudiubelu commented:
> Hello @claudiubelu, my name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Is this still on track for 1.25?

Yes. I've addressed the comments above, waiting for other reviews / for this to get merged.

claudiubelu commented:
/test pull-kubernetes-e2e-gce-storage-slow

claudiubelu commented:
FYI: This PR has minimal code changes, and it mostly contains test changes. This PR has been included in this PR https://github.com/kubernetes/kubernetes/pull/110981, in which we've ran the ``pull-ci-kubernetes-unit-windows`` job, in which we can see that these tests now pass.

claudiubelu commented:
/test pull-kubernetes-e2e-capz-azure-disk

claudiubelu commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

claudiubelu commented:
/assign @thockin

claudiubelu commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

thockin commented on pkg/volume/csi/csi_mounter_test.go:
```diff
@@ -903,19 +903,25 @@ func TestUnmounterTeardown(t *testing.T) {
 	diskMounter := util.NewSafeFormatAndMountFromHost(plug.GetPluginName(), plug.host)
-	if err := diskMounter.FormatAndMount("/fake/device", dir, "testfs", nil); err != nil {
+	device := "/fake/device"
+	if goruntime.GOOS == "windows" {
```
@BenTheElder I know you have feelings about this.  I am inclined to approve, but finding a more consistent pattern would be good.

thockin commented on pkg/volume/testing/testing.go:
```diff
@@ -1146,7 +1146,7 @@ func (fc *FakeProvisioner) Provision(selectedNode *v1.Node, allowedTopologies []
 		}
 	}
-	fullpath := fmt.Sprintf("/tmp/hostpath_pv/%s", uuid.NewUUID())
+	fullpath := fmt.Sprintf("/%s/hostpath_pv/%s", os.TempDir(), uuid.NewUUID())
```
Why does this not need to use Join?  Knowing that code tends to get copied, this should either be consistent or comment why not?

thockin reviewed (commented):
Thanks!

/lgtm
/approve
